storage: add fix and regression test for #7899

tbg opened the pull request:
Fixes #7899 by preventing timestamp cache information loss when
the right hand side lease holder was not equal to that of the
left hand side.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7955)

<!-- Reviewable:end -->

tbg committed:
storage: rename a split test

tbg committed:
roachpb: clarify scan/delete bounds

tbg committed:
testcluster: export lookupRange

tbg committed:
testcluster: wrap some errors

petermattis commented:
Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 2 of 2 files at r8.
Review status: 5 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/client_split_test.go, line 1148 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCf1d1XcsmJuzoe0O8:-KNCf1d22YPyhoj_czRJ:-711446471) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1148)):*
> ```Go
> }
> 
> func addr(k roachpb.Key) roachpb.RKey {
> ```

`s/addr/mustAddr/g`? Or export `keys.mustAddr`.

---

*[storage/client_split_test.go, line 1198 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCfPiFJNN-VsNm0xHy:-KNCfPiGlYLp_JQZlUmY:2073276012) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1198)):*
> ```Go
> 		}
> 	}
> 	if _, err := tc.AddReplicas(addr(leftKey), tc.Target(1)); err != nil {
> ```

Why is `addr` needed here? Perhaps `AddReplicas` should take a `roachpb.Key`.

---

*[storage/client_split_test.go, line 1218 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCg9AyXdFP_JlwJ_EK:-KNCg9AyXdFP_JlwJ_EL:-1291670339) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1218)):*
> ```Go
> 	txnOld := <-txnCh
> 
> 	// Do something with txnOld so that its timestamp gets fixed.
> ```

`s/gets fixed/gets set/g`?

---

*[storage/client_split_test.go, line 1223 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCgF3wmVnrMqc-eSoy:-KNCgF3wmVnrMqc-eSoz:1181575703) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1223)):*
> ```Go
> 	}
> 
> 	// Another client comes along at higher timestamp, touching everything on
> ```

`s/at higher/at a higher/g`

---

*[storage/client_split_test.go, line 1234 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCgNZHnXsiXA84eZGc:-KNCgNZHnXsiXA84eZGd:735767040) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1234)):*
> ```Go
> 	// the other node gets to be the lease holder instead).
> 	// This makes sure that once we split, we'll be in the situation described
> 	// in #7899: The first lease holder of the right hand side of the Split
> ```

I think it is worthwhile to describe the situation in #7899 here instead of referencing that issue. Or maybe better to include the description next to the eventual fix. 

---

*[storage/client_split_test.go, line 1254 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCh1Lfc8x-zAWfsqpf:-KNCh1Lfc8x-zAWfsqpg:-684590519) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1254)):*
> ```Go
> 	// Pull the trigger. This actually also reads the RHS descriptor after the
> 	// split, so when this returns, we've got the leases set up already (not
> 	// that it would matter).
> ```

Nit: add a blank line between paragraphs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'll add a unit test which tests this at a lower level, too. But the bulk here is hungry for review.

---

Review status: 0 of 12 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1148 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCf1d1XcsmJuzoe0O8:-KNCufO19NcFonaFCoZ8:1549158326) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1148)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/addr/mustAddr/g`? Or export `keys.mustAddr`.
</details>
Exported.

---

*[storage/client_split_test.go, line 1198 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCfPiFJNN-VsNm0xHy:-KNCwBzTQnPcjcICwSnv:-418452415) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1198)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Why is `addr` needed here? Perhaps `AddReplicas` should take a `roachpb.Key`.
</details>
Done (with some fanout).

---

*[storage/client_split_test.go, line 1218 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCg9AyXdFP_JlwJ_EK:-KNCxyFQEnm73FhpDBro:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1218)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/gets fixed/gets set/g`?
</details>
Done.

---

*[storage/client_split_test.go, line 1223 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCgF3wmVnrMqc-eSoy:-KNCy4RNuT56s1XG4OdB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1223)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/at higher/at a higher/g`
</details>
Done.

---

*[storage/client_split_test.go, line 1234 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCgNZHnXsiXA84eZGc:-KNCy5m_gc4J037V6POj:-1689333618) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1234)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think it is worthwhile to describe the situation in #7899 here instead of referencing that issue. Or maybe better to include the description next to the eventual fix. 
</details>
Added full description next to fix (and reading along with the test here is hopefully a good summary of what went wrong - if not please add your comments).

---

*[storage/client_split_test.go, line 1254 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNCh1Lfc8x-zAWfsqpf:-KNCyR4h1AIgJ98OfvIS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e8a755ce73ecf53c0953971f4f0f57b31801446/storage/client_split_test.go#L1254)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Nit: add a blank line between paragraphs.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-:-KNEPTWU4Oy4B1eOPqjI:1020532908)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r10, 3 of 3 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 1 of 1 files at r14, 8 of 8 files at r15.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1057 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEqf4jMuQ1Ipitx36I:-KNEqf4jMuQ1Ipitx36J:-1636005731) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d45696ea776a5b912262e5eef9889eacb6abe41/storage/client_split_test.go#L1057)):*
> ```Go
> // TestStoreSplitReadRace prevents regression of #3148. It begins a couple of
> // read requests and lets them complete while a split is happening; the reads
> // hit the second half of the split. If the split happens non-atomically with
> ```

optional: "second" while you're here

---

*[storage/client_split_test.go, line 1152 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEs42m8LUWzyCuAEBu:-KNEs42m8LUWzyCuAEBv:-492604523) ([raw file](https://github.com/cockroachdb/cockroach/blob/598b101d3a93ffb410f1e5c39ab7932f6029619f/storage/client_split_test.go#L1152)):*
> ```Go
> // not equal to the left-hand side lease holder (at the time of the split),
> // its timestamp cache would not be properly initialized, which would allow
> // for writes which violated reads previously served by the pre-split lease.
> ```

nit: not sure "violated" is the right word

---

*[storage/client_split_test.go, line 1199 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEtzuzVWxJXrGU6PtZ:-KNEtzuzVWxJXrGU6Pt_:-1991545911) ([raw file](https://github.com/cockroachdb/cockroach/blob/598b101d3a93ffb410f1e5c39ab7932f6029619f/storage/client_split_test.go#L1199)):*
> ```Go
> 		if err := tc.Stopper().RunAsyncTask(func() {
> 			_ = db.Txn(func(innerTxn *client.Txn) error {
> 				txnCh <- innerTxn // sneakily move txn outside
> ```

why do you need this sneakiness? ISTM that you can use client.NewTxn

---

*[storage/client_split_test.go, line 1256 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEtcv7htBUctHiMdl0:-KNEtcv7htBUctHiMdl1:-1681246454) ([raw file](https://github.com/cockroachdb/cockroach/blob/598b101d3a93ffb410f1e5c39ab7932f6029619f/storage/client_split_test.go#L1256)):*
> ```Go
> 	// cache which does not exhibit the anomaly.
> 	//
> 	// In practice, this should only be possible if second-long delays occur
> ```

we could make this more robust by increasing the leader lease duration for this test, right? perhaps a TODO that we should add that to the testing hooks.

---

*[storage/replica_command.go, line 2515 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNErMRK0awbhReiZKBm:-KNErMRK0awbhReiZKBn:1347334933) ([raw file](https://github.com/cockroachdb/cockroach/blob/973d45ad0b1d5f170de33b7d1fb5a6b84cc62bfe/storage/replica_command.go#L2515)):*
> ```Go
> 	// - node b becomes the lease holder for [c,e). Its timestamp cache does
> 	//   know about the read at 'd' which happened at the beginning.
> 	// - node b can illegaly propose a write to 'd' for a lower timestamp.
> ```

s/for/at/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

as usual, i'd prefer to see the test and the fix in the same commit.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-:-KNEuAzeaW4H5BdYcB7I:-1650211622)*
<!-- Sent from Reviewable.io -->

tbg committed:
stop: add WithCancel

tschottdorf commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1057 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEqf4jMuQ1Ipitx36I:-KNGTDs8P3Q_99CzF9L2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d45696ea776a5b912262e5eef9889eacb6abe41/storage/client_split_test.go#L1057)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> optional: "second" while you're here
</details>
Done.

---

*[storage/client_split_test.go, line 1152 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEs42m8LUWzyCuAEBu:-KNGTLRDz3y_1TF8DgV2:-1266068883) ([raw file](https://github.com/cockroachdb/cockroach/blob/598b101d3a93ffb410f1e5c39ab7932f6029619f/storage/client_split_test.go#L1152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: not sure "violated" is the right word
</details>
Done (`invalidated`).

---

*[storage/client_split_test.go, line 1199 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEtzuzVWxJXrGU6PtZ:-KNGTafvnlWyKsnz73JJ:503908658) ([raw file](https://github.com/cockroachdb/cockroach/blob/598b101d3a93ffb410f1e5c39ab7932f6029619f/storage/client_split_test.go#L1199)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why do you need this sneakiness? ISTM that you can use client.NewTxn
</details>
I can't argue with that. Done.

---

*[storage/client_split_test.go, line 1256 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNEtcv7htBUctHiMdl0:-KNH9N1lH8F6a4aEWNYp:329497063) ([raw file](https://github.com/cockroachdb/cockroach/blob/598b101d3a93ffb410f1e5c39ab7932f6029619f/storage/client_split_test.go#L1256)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> we could make this more robust by increasing the leader lease duration for this test, right? perhaps a TODO that we should add that to the testing hooks.
</details>
Added an assertion below. The concern here is purely hypothetical.

---

*[storage/replica_command.go, line 2515 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNErMRK0awbhReiZKBm:-KNH9UMEAcnVsqY_Dyps:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/973d45ad0b1d5f170de33b7d1fb5a6b84cc62bfe/storage/replica_command.go#L2515)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/for/at/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r17, 2 of 2 files at r20, 9 of 9 files at r21.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 3 of 3 files at r11, 1 of 1 files at r14, 6 of 8 files at r15, 2 of 3 files at r17, 2 of 2 files at r20, 9 of 9 files at r21.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 2513 \[r21\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNHpAocpAvQ7yrbj9nf:-KNHpAod4AnnAEJYMmMZ:619868368) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee9a71ce5deb4383bf88d8bfb707644901f5030a/storage/replica_command.go#L2513)):*
> ```Go
> 	//   Replica. However, only node one's timestamp cache contains information
> 	//   about the read of key 'd' in the first place.
> 	// - node b becomes the lease holder for [c,e). Its timestamp cache does
> ```

I guess you mean "node two" here not "b".

---

*[storage/replica_command.go, line 2515 \[r21\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNHpDlCXkWU1mIqSIIi:-KNHpDlCXkWU1mIqSIIj:678021867) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee9a71ce5deb4383bf88d8bfb707644901f5030a/storage/replica_command.go#L2515)):*
> ```Go
> 	// - node b becomes the lease holder for [c,e). Its timestamp cache does
> 	//   know about the read at 'd' which happened at the beginning.
> 	// - node b can illegaly propose a write to 'd' at a lower timestamp.
> ```

same

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: maintain lease holder on split

Fixes #7899 by preventing timestamp cache information loss when
the right hand side lease holder was not equal to that of the
left hand side.

Added a relatively high-level regression test which mirrors the
anomaly from the client's perspective, as observed in Jepsen testing,
and also checks that the test passing is due to the fix - initializing
the right-hand side with the lease of the left-hand side.

tschottdorf commented:
Review status: 13 of 14 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/replica_command.go, line 2513 \[r21\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNHpAocpAvQ7yrbj9nf:-KNHqN6EGBrWipPdRgpc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee9a71ce5deb4383bf88d8bfb707644901f5030a/storage/replica_command.go#L2513)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I guess you mean "node two" here not "b".
</details>
Done.

---

*[storage/replica_command.go, line 2515 \[r21\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-KNHpDlCXkWU1mIqSIIi:-KNHqNNziNEEKg7_o5Vi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee9a71ce5deb4383bf88d8bfb707644901f5030a/storage/replica_command.go#L2515)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> same
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 1 files at r22.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-:-KNI-3viSrFhIwcs0CXE:1417088721)*
<!-- Sent from Reviewable.io -->

tbg committed:
server: fix buglet in TestRangesResponse

The test was verifying that a single node cluster has only Raft leaders, but
was not taking into account the possibility of a dormant replica (which has
been made far more likely ever since reads to the newly created right-hand
side of the split do not have to acquire a lease, and thus do not wake up
Raft).

Fixes #7977.

tschottdorf commented:
Fixed a buglet, PTAL.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 14 of 16 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955#-:-KNI8Yz0EqSuyt7RGDMd:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r22, 2 of 2 files at r23.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7955)*
<!-- Sent from Reviewable.io -->
